### PR TITLE
React basic practice - develop - Fix/product item - Update data / Add item to cart #80

### PR DESCRIPTION
**Pull request by minhng271**
### Issue
- Update field data
- Check quantity -> show label sold out / discount
- Limited quantity added
- Wrap tags for images and product names
### Result

https://user-images.githubusercontent.com/49194444/217413880-67124e42-c6d8-4ef9-819f-c95590a41e21.mp4

### CODE REVIEW

File: react-basic-practice/src/components/ProductCard/index.jsx [hunk 1]
```diff
@@ -1,18 +1,31 @@
...
 
 const ProductCard = (props) => {
   const { product } = props;
-  const price = product.price.toFixed(2);
+  let label = '';
+
+  const showLabel = () => {
+    let haveLabel = false;
+    if (!product.quantity) {
+      label = 'Sold out';
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/components/ProductCard/index.jsx]: you should not have side effect here. This needs to be a pure function. Please google it.

File: react-basic-practice/src/pages/ProductDetail/index.jsx [hunk 2]
```diff
@@ -65,32 +65,51 @@ const ProductDetail = () => {
...
+                name: product.name,
+                price: setPrice(),
+                discountPercent: product.discountPercent,
+              },
+            ],
+          };
+        }
+
+        setCart(cartUser);
+        await updateCart(userData.userId, cartUser);
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/pages/ProductDetail/index.jsx:100]: You should have loading icon when updating cart. Please follow the example app.

File: react-basic-practice/src/pages/ProductDetail/index.jsx [hunk 3]
```diff
@@ -65,32 +65,51 @@ const ProductDetail = () => {
...
+
+        setToast({
+          openPopup: true,
+          status: 'success',
+          message: 'The item added to your shopping bag',
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/pages/ProductDetail/index.jsx]: The item has been added to your shopping bag.

File: react-basic-practice/src/components/ProductCard/index.jsx [hunk 4]
```diff
@@ -1,18 +1,27 @@
...
+import { Link } from 'react-router-dom';
 import './index.css';
 
 const ProductCard = (props) => {
   const { product } = props;
-  const price = product.price.toFixed(2);
+
+  const renderLabel = (quantity, discountPercent) => {
+    if (quantity === 0) {
+      return <span className="product-label">Sold out</span>;
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/components/ProductCard/index.jsx]: You just need to return the string.

File: react-basic-practice/src/components/ProductCard/index.jsx [hunk 5]
```diff
@@ -1,18 +1,27 @@
 import React from 'react';
+import { Link } from 'react-router-dom';
 import './index.css';
 
 const ProductCard = (props) => {
   const { product } = props;
-  const price = product.price.toFixed(2);
+
+  const renderLabel = (quantity, discountPercent) => {
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/components/ProductCard/index.jsx:16]: Methods must have comment.

File: react-basic-practice/src/components/Quantity/index.jsx [hunk 6]
```diff
@@ -9,12 +9,14 @@ import Input from '../common/Input';
 import './index.css';
 
 const Quantity = (props) => {
-  const { isUnavailableProduct, onChangeQuantity } = props;
+  const { isUnavailableProduct, onChangeQuantity, limitQuantity } = props;
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/components/Quantity/index.jsx]: limitQuantity  => maxQuantity

File: react-basic-practice/src/components/common/Input/index.jsx [hunk 7]
```diff
@@ -26,25 +27,20 @@ const Input = (props) => {
...
-        {label}
-      </label>
+      <label htmlFor={htmlFor}>{label}</label>
       <input
-        type={inputType}
+        type={type}
         name={name}
         className={`input-text ${className}`}
         onChange={handleChange}
+        onInput={handleInput}
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/components/common/Input/index.jsx]: onInput what?

Comment (minhng271) [reply react-basic-practice/src/components/common/Input/index.jsx]: To block users from entering negative numbers (catch events as soon as input)

File: react-basic-practice/src/components/ProductCard/index.jsx (hunk 5 above)
Comment (MinhNguyen41092) [reply react-basic-practice/src/components/ProductCard/index.jsx:16]: Must specify data type of params.

File: react-basic-practice/src/components/ProductCard/index.jsx [hunk 8]
```diff
@@ -1,18 +1,30 @@
...
 
   return (
     <article className={`product-card ${product.id}`}>
-      {
-        product.label && <span className="product-label">{product.label}</span>
-      }
-      <img className="img" src={product.image} alt="product" />
-      <h4 className="name">{product.name}</h4>
-      <span className="price">{`$ ${price}`}</span>
+      {(product.quantity <= 0 || product.discountPercent > 0) &&
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/components/ProductCard/index.jsx]: Perform this check inside renderLabel.

File: react-basic-practice/src/components/common/Input/index.jsx (hunk 7 above)
Comment (MinhNguyen41092) [reply react-basic-practice/src/components/common/Input/index.jsx]: Do you think the name `onInput` reflect that?

Comment (MinhNguyen41092) [reply react-basic-practice/src/components/common/Input/index.jsx]: What input event does it handle?

File: react-basic-practice/src/pages/ProductDetail/index.jsx [hunk 9]
```diff
@@ -55,42 +55,63 @@ const ProductDetail = () => {
...
       const indexProduct = cart?.products?.findIndex(
-        (item) => item.idProduct === Number(productId),
+        (item) => item.idProduct === Number(productId)
       );
 
-      if (indexProduct >= 0) {
-        cart.products[indexProduct].quantity += quantityProduct;
-        cartUser = cart;
+      setLoading(true);
+      if (quantityProduct && quantityProduct < product.quantity) {
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/pages/ProductDetail/index.jsx]: What is `quantityProduct`?

File: react-basic-practice/src/pages/ProductDetail/index.jsx [hunk 10]
```diff
@@ -55,42 +55,63 @@ const ProductDetail = () => {
...
+        if (indexProduct >= 0) {
+          cart.products[indexProduct].quantity += quantityProduct;
+          cartUser = cart;
+        } else {
+          const setPrice = () => {
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/pages/ProductDetail/index.jsx]: Move this outside.

File: react-basic-practice/src/components/common/Input/index.jsx (hunk 7 above)
Comment (minhng271) [reply react-basic-practice/src/components/common/Input/index.jsx]: `onInput` is an attribute of input. The `oninput` attribute fires when an element gets user input.

File: react-basic-practice/src/components/ProductCard/index.jsx [hunk 11]
```diff
@@ -1,18 +1,35 @@
...
+   *
+   * @param {*} quantity available of product
+   * @param {*} discountPercent of product
+   *
+   * @returns component label
+   */
+  const renderLabel = (quantity, discountPercent) => {
+    let label = '';
+    if (product.quantity <= 0 || product.discountPercent > 0) {
+      quantity === 0 ? (label = 'Sold out') : (label = `${discountPercent} %`);
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/components/ProductCard/index.jsx]: label = quantity === 0 ? 'sold out' :  `${discountPercent} %`

**Review by MinhNguyen41092 — APPROVED** [review]

**Pull request merged**